platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java

```diff
@@ -1681,10 +1682,32 @@ public void addActionMethod(Method method) {
             System.arraycopy(originalParameterTypes, 0, closurisedParameterTypes, 0, numParams);
             closurisedParameterTypes[numParams - 1] = CLOSURE_TYPE;
 
+            final String actionGenericTypeName;
+            java.lang.reflect.Type actionParameterType = method.getGenericParameterTypes()[numParams - 1];
+            if (actionParameterType instanceof ParameterizedType) {
+                ParameterizedType actionParameter = (ParameterizedType) actionParameterType;
+                java.lang.reflect.Type actionGenericType = actionParameter.getActualTypeArguments()[0];
+
+                if (actionGenericType instanceof WildcardType) {
+                    WildcardType wildcardType = (WildcardType) actionGenericType;
+                    java.lang.reflect.Type[] upperBounds = wildcardType.getLowerBounds();
+                    actionGenericTypeName = upperBounds[0].getTypeName();
+                } else {
+                    actionGenericTypeName = actionGenericType.getTypeName();
+                }
+            } else {
+                actionGenericTypeName = null;
+            }
+
             final String methodDescriptor = getMethodDescriptor(returnType, closurisedParameterTypes);
 
-            // GENERATE public <return type> <method>(Closure v) { return <method>(…, ConfigureUtil.configureUsing(v)); }
+            // GENERATE public <return actionParamter> <method>(Closure v) { return <method>(…, ConfigureUtil.configureUsing(v)); }
             publicMethod(method.getName(), methodDescriptor, methodVisitor -> new MethodVisitorScope(methodVisitor) {{
+                if (actionGenericTypeName != null) {
+                    AnnotationVisitor annotationVisitor = visitParameterAnnotation(numParams - 1, "Lgroovy/lang/DelegatesTo;", true);
+                    annotationVisitor.visitAnnotation("value", "L" + actionGenericTypeName.replace('.', '/') + ";");
```

Comment:
Some of the weirdness might be this line -- changing it to
```java
annotationVisitor.visit("value", Type.getObjectType(actionGenericTypeName.replace('.', '/')));
```
Could fix that, since `DelegatesTo#value` is a `Class`, not a nested annotation.